source/modulo_components/include/modulo_components/ComponentInterface.hpp

```diff
@@ -425,6 +455,98 @@ void ComponentInterface<NodeT>::set_predicate(
+    auto message_pair = make_shared_message_pair(data, this->get_clock());
+    std::shared_ptr<SubscriptionInterface> subscription_interface;
+    switch (message_pair->get_type()) {
+      case MessageType::BOOL: {
+        auto subscription_handler = std::make_shared<SubscriptionHandler<std_msgs::msg::Bool>>(message_pair);
+        auto subscription = NodeT::template create_subscription<std_msgs::msg::Bool>(
+            topic_name, this->qos_, subscription_handler->get_callback());
+        subscription_interface = subscription_handler->create_subscription_interface(subscription);
+        break;
+      }
+      case MessageType::FLOAT64: {
+        auto subscription_handler = std::make_shared<SubscriptionHandler<std_msgs::msg::Float64>>(message_pair);
+        auto subscription = NodeT::template create_subscription<std_msgs::msg::Float64>(
+            topic_name, this->qos_, subscription_handler->get_callback());
+        subscription_interface = subscription_handler->create_subscription_interface(subscription);
+        break;
+      }
+      case MessageType::FLOAT64_MULTI_ARRAY: {
+        auto subscription_handler =
+            std::make_shared<SubscriptionHandler<std_msgs::msg::Float64MultiArray>>(message_pair);
+        auto subscription = NodeT::template create_subscription<std_msgs::msg::Float64MultiArray>(
+            topic_name, this->qos_, subscription_handler->get_callback());
+        subscription_interface = subscription_handler->create_subscription_interface(subscription);
+        break;
+      }
+      case MessageType::INT32: {
+        auto subscription_handler = std::make_shared<SubscriptionHandler<std_msgs::msg::Int32>>(message_pair);
+        auto subscription = NodeT::template create_subscription<std_msgs::msg::Int32>(
+            topic_name, this->qos_, subscription_handler->get_callback());
+        subscription_interface = subscription_handler->create_subscription_interface(subscription);
+        break;
+      }
+      case MessageType::STRING: {
+        auto subscription_handler = std::make_shared<SubscriptionHandler<std_msgs::msg::String>>(message_pair);
+        auto subscription = NodeT::template create_subscription<std_msgs::msg::String>(
+            topic_name, this->qos_, subscription_handler->get_callback());
+        subscription_interface = subscription_handler->create_subscription_interface(subscription);
+        break;
+      }
+      case MessageType::ENCODED_STATE: {
+        auto subscription_handler = std::make_shared<SubscriptionHandler<modulo_new_core::EncodedState>>(message_pair);
+        auto subscription = NodeT::template create_subscription<modulo_new_core::EncodedState>(
+            topic_name, this->qos_, subscription_handler->get_callback());
+        subscription_interface = subscription_handler->create_subscription_interface(subscription);
+        break;
+      }
+    }
+    this->inputs_.insert(std::make_pair(parsed_signal_name, subscription_interface));
```

Comment:
I didn't think about this in the publisher, but we should check if the input / output already exists before doing all of this work. If the signal name already exists, `insert` will not change the map, but the subscription will have been created. We shouldn't allow creating two inputs with the same signal name, because ROS gets particular about creating and registering the same pubs / subs multiple times. That should be checked in the inputs_ map at the start of this function, before any of the subsequent logic. 

As a side note, I don't really like `insert(std::make_pair(..., ...))` and would prefer the more concise member initialiser list, available since C++11, as:
```suggestion
    this->inputs_.insert({parsed_signal_name, subscription_interface});
```
I know it's done the other way through the rest of the file so you don't have to change it, just pointing it out. And, where appropriate, the C++17 extensions `try_emplace` and `insert_or_assign` are preferable still. If you check for prior signal name existence in the map beforehand anyway, you could directly use `insert_or_assign(signal_name, interface)`.